webapp/src/main/java/com/box/l10n/mojito/android/strings/AndroidStringDocumentMapper.java

```diff
@@ -0,0 +1,190 @@
+                        builder.setName(textUnit.getAssetPath() + assetDelimiter + name);
+                        builder.setComment(textUnit.getComment());
+                    }
+
+                    builder.addItem(new AndroidPluralItem(
+                            textUnit.getPluralForm(),
+                            textUnit.getTmTextUnitId(),
+                            useSource ? textUnit.getSource() : textUnit.getTarget()));
+
+                    return builder;
+
+                });
+            }
+        });
+
+        pluralByOther.forEach((pluralFormOther, builder) -> document.addPlural(builder.build()));
+
+        return document;
+    }
+
+    public AndroidStringDocument readFromSourceTextUnits(List<TextUnitDTO> textUnits) {
+        return readFromTextUnits(textUnits, true);
+    }
+
+    public AndroidStringDocument readFromTargetTextUnits(List<TextUnitDTO> textUnits) {
+        return readFromTextUnits(textUnits, false);
+    }
+
+    public List<TextUnitDTO> mapToTextUnits(AndroidStringDocument document) {
+        return mapToTextUnitStream(document).collect(Collectors.toList());
+    }
+
+    public Stream<TextUnitDTO> mapToTextUnitStream(AndroidStringDocument document) {
+        return document.getStrings()
+                .stream()
+                .flatMap(this::stringToTextUnits)
+                .map(this::addTextUnitDTOAttributes);
+    }
+
+    TextUnitDTO addTextUnitDTOAttributes(TextUnitDTO textUnit) {
+
+        if (!Strings.isNullOrEmpty(locale)){
+            textUnit.setTargetLocale(locale);
+        }
+
+        if (!Strings.isNullOrEmpty(repositoryName)){
+            textUnit.setRepositoryName(repositoryName);
+        }
+
+        if (textUnit.getName().contains(assetDelimiter)){
+            String[] nameParts = textUnit.getName().split(assetDelimiter);
+
+            if (nameParts.length == 2) {
+                textUnit.setAssetPath(nameParts[0]);
+                textUnit.setName(nameParts[1]);
+            }
+        }
+
+        return textUnit;
+    }
+
+    Stream<TextUnitDTO> stringToTextUnits(AbstractAndroidString androidString) {
+        Stream<TextUnitDTO> result;
+
+        if (androidString.isSingular()){
+            result = singularToTextUnit((AndroidSingular) androidString);
+        } else {
+            result = pluralToTextUnits((AndroidPlural) androidString);
+        }
+
+        return result;
+    }
+
+    Stream<TextUnitDTO> singularToTextUnit(AndroidSingular singular) {
+        TextUnitDTO textUnit = new TextUnitDTO();
+
+        textUnit.setName(singular.getName());
+        textUnit.setComment(singular.getComment());
+        textUnit.setTmTextUnitId(singular.getId());
+        textUnit.setTarget(unescape(singular.getContent()));
+
+        return Stream.of(textUnit);
+    }
+
+    Stream<TextUnitDTO> pluralToTextUnits(AndroidPlural plural) {
+        Stream.Builder<TextUnitDTO> builder = Stream.builder();
```

Comment:
what's the point of the builder here vs. something like "`plural.items.stream().map(item -> {.... return textUnitDTO})`" ?

Reply:
Updated.